README.md
# StealTheTrophy
My game


pig-pursuit-master/changelog.txt
changelog:

9.29.2021:
changed textures & sounds

9.30.2021
changed textures & sounds
added "bucket button"
fixed "reset game" button
changed firing speed

10.1.2021
nothing

10.2.2021
changed textures
rewrote level 2 partially

10.3.2021
rewrote level 2 partially

10.4.2021
rewrote level 2 partially

10.15.2021
rewrote level 1 partially

10.17.2021
rewrote level 1 partially
added KillCount

10.18.2021
finished level 1 rewrite
added trophy and added it to end of level 2
changed bullet color

10.20.2021
Finished!




pig-pursuit-master/js/pig.core.js
/**
 * Pig Pursuit
 *
 * @fileoverview Core game functionality.
 * @link https://github.com/stevenbenner/pig-pursuit
 * @author Steven Benner
 * @version 0
 * @license Pig Pursuit
 * <https://github.com/stevenbenner/pig-pursuit>
 * Copyright (c) 2012 Steven Benner, http://stevenbenner.com/
 */

pig.core = {};

pig.core.init = function(level) {
	// create session
	pig.session.level = level;
	pig.session.hero = new pig.objects.hero(pig.gfx.sprites.heroLeft, pig.gfx.sprites.heroRight);
	pig.session.level.init();

	// create the canvas
	pig.gfx.createCanvas();

	pig.core.resetGame();

	// load background images
	pig.gfx.loadImages(function() {
		pig.audio.loadAudio(pig.core.execMainGameLoop);
	});


	var start = new Date();
	setInterval(function() {
		var now = new Date(),
			delta = now - start;
		pig.session.fps = pig.session.frames / delta * 1000;
		pig.session.frames = 0;
		start = now;
	}, 250);

	// attach input listeners
	addEventListener('keydown', function(event) {
		if (event.keyCode === 32 || event.keyCode === 37 || event.keyCode === 39) {
			event.preventDefault();
		}
		if (!pig.session.level.isCutscene) {
			pig.session.keysDown[event.keyCode] = true;
		}
	}, false);

	addEventListener('keyup', function(event) {
		delete pig.session.keysDown[event.keyCode];
	}, false);

	addEventListener('mousemove', function(event) {
		if (!pig.session.level.isCutscene) {
			pig.session.mouseX = event.pageX;
			pig.session.mouseY = event.pageY;
		}
	});

	addEventListener('mousedown', function(event) {
		if (!pig.session.level.isCutscene) {
			if (event.button !== 2 && !pig.session.dead && !pig.session.paused) {
				var canvasPosX = pig.session.canvas.offsetLeft,
					canvasPosY = pig.session.canvas.offsetTop,
					heroRealX = canvasPosX + pig.session.hero.x + 13 + pig.session.xOffset,
					heroRealY = canvasPosY + pig.session.hero.y + 19;

				pig.session.projectiles.push(
					new pig.objects.projectile(
						pig.session.hero.x + 13,
						pig.session.hero.y + 19,
						Math.atan2(event.pageY - heroRealY, event.pageX - heroRealX)
					)
				);

				pig.audio.play('shoot');
			}
		}
	}, false);

};

pig.core.execMainGameLoop = function() {
	var lastTime = Date.now();

	pig.session.levelStartTime = lastTime;

	var main = function() {
		var now = Date.now(),
			delta = now - lastTime;

		pig.core.updateGame(delta / 1000);
		pig.gfx.renderFrame();

		pig.session.frames++;

		lastTime = now;

		if (!pig.session.dead && !pig.session.paused) {
			window.requestAnimFrame(main);
		} else {
			pig.session.shootInterval = clearInterval(pig.session.shootInterval);
			//pig.audio.pause('bgm');
		}
	};

	pig.session.shootInterval = setInterval(function() {
		var i;
		if (!pig.session.level.isCutscene) {
			for (i = pig.session.level.monsters.length; i--;) {
				if (Math.abs(pig.session.level.monsters[i].x - pig.session.hero.x) < 512 &&
					Math.abs(pig.session.level.monsters[i].y - pig.session.hero.y) < 256 &&
					pig.session.level.monsters[i].x + pig.session.xOffset > 0 &&
					pig.session.level.monsters[i].x + pig.session.xOffset < pig.session.canvas.width) {
					pig.session.projectiles.push(
						new pig.objects.projectile(
							pig.session.level.monsters[i].x + 13,
							pig.session.level.monsters[i].y + 13,
							Math.atan2(
								pig.session.hero.y - pig.session.level.monsters[i].y,
								pig.session.hero.x - pig.session.level.monsters[i].x
							),
							128
						)
					);
					pig.audio.play('shoot2');
				}
			}
		}
	}, 250);

	//pig.audio.play('bgm');

	window.requestAnimFrame(main);
};

pig.core.pause = function() {
	if (pig.session.paused) {
		pig.session.paused = false;
		pig.core.execMainGameLoop();
	} else {
		pig.session.shootInterval = clearInterval(pig.session.shootInterval);
		pig.session.paused = true;
	}
};

pig.core.resetGame = function() {
	pig.session.paused = false;
	pig.session.dead = false;
	pig.session.xOffset = 0;
	pig.session.keysDown = {};
	pig.session.score = 0;
	pig.session.projectiles = [];
	pig.session.hero.health = 256;
	pig.session.hero.x = pig.session.level.spawnX;
	pig.session.hero.y = pig.session.level.spawnY;
};

pig.core.updateGame = function(modifier) {
	pig.core.executeEvents(modifier);

	pig.core.updateHero(modifier);
	pig.core.updateProjectiles(modifier);
	pig.core.updateMonsters(modifier);
};

var activeEvents = [];

pig.core.executeEvents = function(modifier) {
	var now = Date.now(),
		currentTime = now - pig.session.levelStartTime,
		lastTime = currentTime - (modifier*1000),
		i, distance;
	// find events that need to start
	for (i = pig.session.level.events.length; i--;) {
		if (pig.session.level.events[i].startTime >= lastTime && pig.session.level.events[i].startTime <= currentTime) {
			pig.session.level.events[i].startTime = now;
			activeEvents.push(pig.session.level.events[i]);
		}
	}
	// do events
	for (i = activeEvents.length; i--;) {
		if (activeEvents[i].duration >= 0) {
			if (activeEvents[i].type === 'move') {
				distance = activeEvents[i].action * modifier;
				activeEvents[i].object.x += distance;
				pig.session.mouseX = 10000 * activeEvents[i].action;
			} else if (activeEvents[i].type === 'say') {
				activeEvents[i].object.message = activeEvents[i].action;
			}
			activeEvents[i].duration -= (modifier*1000);
		} else {
			if (activeEvents[i].type === 'say') {
				activeEvents[i].object.message = null;
			}
			activeEvents.splice(i, 1);
		}
	}
};

pig.core.updateHero = function(modifier) {
	var pixelsTraverced = pig.session.hero.speed * modifier;

	// left arrow or 'w'
	if (37 in pig.session.keysDown || 65 in pig.session.keysDown) { 
		if (!hasLeftCollision(pig.session.hero)) {
			// move hero
			pig.session.hero.x -= pixelsTraverced;
			//pig.session.hero.x = (0.5 + pig.session.hero.x) << 0; // fast round
			// move viewport
			if (pig.session.hero.x + pig.session.xOffset < 300 && pig.session.xOffset < 0) {
				pig.session.xOffset += pixelsTraverced;
				pig.session.xOffset = Math.round(pig.session.xOffset); //(0.5 + pig.session.xOffset) << 0; // fast round
				if (pig.session.xOffset > 0) {
					pig.session.xOffset = 0;
				}
			}
		} else if (hasLeftCollision(pig.session.hero).type === 'door') {
			var nextLevel = new pig.objects.level2();
			nextLevel.imageObject = new Image();
			nextLevel.imageObject.addEventListener('load', (function(nextLevel) {
				return function() {
					pig.session.hero.x = pig.session.canvas.width / 2;
					pig.session.hero.y = pig.session.canvas.height / 2;
					pig.session.hero.health = 256;
					pig.session.xOffset = 0;
					pig.session.level = nextLevel;

		

					
				}
			})(nextLevel), false);
			nextLevel.imageObject.src = nextLevel.backgroundImage;
		} else if (hasLeftCollision(pig.session.hero).type === 'tropyh') {
			var nextLevel = new pig.objects.level3();
			nextLevel.imageObject = new Image();
			nextLevel.imageObject.addEventListener('load', (function(nextLevel) {
				return function() {
					pig.session.hero.x = pig.session.canvas.width / 2;
					pig.session.hero.y = pig.session.canvas.height / 2;
					pig.session.hero.health = 256;
					pig.session.xOffset = 0;
					pig.session.level = nextLevel;
				}
			})(nextLevel), false);
			nextLevel.imageObject.src = nextLevel.backgroundImage;
			
			
		}
	}

	// right arrow or 'd'
	if (39 in pig.session.keysDown || 68 in pig.session.keysDown) { 
		if (!hasRightCollision(pig.session.hero)) {
			// move hero
			pig.session.hero.x += pixelsTraverced;
			//pig.session.hero.x = (0.5 + pig.session.hero.x) << 0; // fast round
			// move viewport
			if (pig.session.hero.x + pig.session.xOffset > pig.session.canvas.width - 300 &&
				pig.session.xOffset > pig.session.canvas.width + pig.session.level.width * -32) {
				pig.session.xOffset -= pixelsTraverced;
				pig.session.xOffset = Math.round(pig.session.xOffset); //(0.5 + pig.session.xOffset) << 0; // fast round
				if (pig.session.xOffset < pig.session.canvas.width + 1 + pig.session.level.width * -32) {
					pig.session.xOffset = pig.session.canvas.width + 1 + pig.session.level.width * -32;
				}
			}
		}
	}

	// space bar
	if (32 in pig.session.keysDown) {
		if (!pig.session.hero.jumping && hasGoundCollision(pig.session.hero)) {
			pig.session.hero.jumping = true;
			pig.session.hero.jumpVelocity = pig.session.hero.speed * 3;
			pig.audio.play('jump');
		}
	}

	// jumping
	if (pig.session.hero.jumping) {
		pig.session.hero.jumpVelocity -= pixelsTraverced * 4;
		if (pig.session.hero.jumpVelocity > 0 && !hasUpCollision()) {
			jump(pig.session.hero, modifier);
		} else {
			pig.session.hero.jumping = false;
		}
	}

	// ground collision
	if (!hasGoundCollision(pig.session.hero) && !pig.session.hero.jumping) {
		fall(pig.session.hero, modifier);
		// if hero fell into a pit
		if (pig.session.hero.y > pig.session.canvas.height) {
			pig.session.dead = true;
			pig.audio.play('die');
		}
	}
};

pig.core.updateProjectiles = function(modifier) {
	var i, n, projectileDistance, newX, newY;

	for (i = pig.session.projectiles.length; i--;) {

		// maths
		projectileDistance = pig.session.projectiles[i].speed * modifier;
		newX = pig.session.projectiles[i].x + projectileDistance * Math.cos(pig.session.projectiles[i].angle);
		newY = pig.session.projectiles[i].y + projectileDistance * Math.sin(pig.session.projectiles[i].angle);

		// set coordinates
		pig.session.projectiles[i].x = newX;
		pig.session.projectiles[i].y = newY;

		// level collisions
		if (getLevelCollision(pig.session.projectiles[i])) {
			//getLevelCollision(pig.session.projectiles[i]).destroy();
			pig.session.projectiles.splice(i, 1);
			continue;
		}

		// monster/hero collisions
		if (pig.session.projectiles[i].isHero) {
			for (n = pig.session.level.monsters.length; n--;) {
				if (Math.abs(pig.session.level.monsters[n].x + 23 - pig.session.projectiles[i].x) < 23 &&
					Math.abs(pig.session.level.monsters[n].y + 23 - pig.session.projectiles[i].y) < 23) {
					pig.session.projectiles.splice(i, 1);
					pig.session.level.monsters[n].health -= 23;
					// if moster is dead
					if (pig.session.level.monsters[n].health <= 0) {
						pig.session.level.monsters.splice(n, 1);
						pig.session.score += 1;
					}
					break;
				}
			}
		} else {
			if (Math.abs(pig.session.hero.x + 23 - pig.session.projectiles[i].x) < 23 &&
				Math.abs(pig.session.hero.y + 23 - pig.session.projectiles[i].y) < 23) {
				pig.session.projectiles.splice(i, 1);
				pig.session.hero.health -= 18;
				// if hero is dead
				if (pig.session.hero.health <= 0) {
					pig.session.dead = true;
					pig.audio.play('die');
				}
			}
		}
		
		// projectile hit monster or hero and was removed
		if (!pig.session.projectiles[i]) {
			continue;
		}

		// destroy off-screen projectiles
		if (pig.session.projectiles[i].x < pig.session.xOffset * -1 ||
			pig.session.projectiles[i].x > pig.session.xOffset * -1 + pig.session.canvas.width) {
			pig.session.projectiles.splice(i, 1);
		}

	}
};

pig.core.updateMonsters = function(modifier) {
	var i;
	for (i = pig.session.level.monsters.length; i--;) {
		// gravity
		if (pig.session.level.monsters[i].type === 'walker' && !hasGoundCollision(pig.session.level.monsters[i])) {
			fall(pig.session.level.monsters[i], modifier);
			// if monster fell into a pit
			if (pig.session.level.monsters[i].y > pig.session.canvas.height) {
				pig.session.level.monsters.splice(i, 1);
			}
		}
		// walkers
		if (pig.session.level.monsters[i].type === 'walker' || pig.session.level.monsters[i].type === 'flyer') {
			if (pig.session.level.monsters[i].east) {
				pig.session.level.monsters[i].x += pig.session.level.monsters[i].speed * modifier;
				if (hasRightCollision(pig.session.level.monsters[i])) {
					pig.session.level.monsters[i].east = false;
				}
			} else {
				pig.session.level.monsters[i].x -= pig.session.level.monsters[i].speed * modifier;			
				if (hasLeftCollision(pig.session.level.monsters[i])) {
					pig.session.level.monsters[i].east = true;
				}
			}
		}
	}
};

function fall(object, modifier) {
	var gridPos = getLevelGridAtCoord(object.x + 16, object.y + 16),
		floorY = 600,
		canvasRowCount = pig.session.canvas.width / GRID_SIZE,
		y;
	for (y = gridPos.y; y < canvasRowCount; y++) {
		if (pig.session.level.content[gridPos.x][y]) {
			floorY = y * GRID_SIZE;
			break;
		}
	}
	if (object.y < floorY - GRID_SIZE) { // if object is above floor
		object.y += 400 * modifier;
		//object.y = (0.5 + object.y) << 0; // fast round
		if (object.y > floorY - GRID_SIZE) {
			object.y = floorY - GRID_SIZE;
		}
	}
}

function jump(object, modifier) {
	var gridPos = getLevelGridAtCoord(object.x + 16, object.y + 16),
		roofY = -128,
		canvasRowCount = pig.session.canvas.width / GRID_SIZE,
		y;
	for (y = gridPos.y; y > 0; y--) {
		if (pig.session.level.content[gridPos.x][y]) {
			roofY = y * 32;
			break;
		}
	}
	if (object.y > roofY + 32) { // if object is below roof
		object.y -= (object.jumpVelocity * modifier) / 2;
		//object.y = (0.5 + object.y) << 0; // fast round
		if (object.y < roofY + 32) {
			object.y = roofY + 32;
			object.jumping = false;
		}
	}
}

function getLevelCollision(object) {
	var grid = getLevelGridAtCoord(object.x, object.y);
	try {
		return pig.session.level.content[grid.x][grid.y];
	} catch(e) {
		return null;
	}
}

function getLevelGridAtCoord(x, y) {
	return {
		x: Math.ceil(x / GRID_SIZE) - 1,
		y: Math.ceil(y / GRID_SIZE) - 1
	};
}


function hasGoundCollision(object) {
	var playerRow = Math.round(object.x / 32),
		playerCol = Math.round((object.y + 16) / 32);
	return pig.session.level.content[playerRow][playerCol];
}
function hasUpCollision() {
	var playerRow = Math.round((pig.session.hero.x) / 32),
		playerCol = Math.round(pig.session.hero.y / 32);
	return pig.session.level.content[playerRow][playerCol];
}
function hasLeftCollision(object) {
	var playerRow = Math.round((object.x - 16) / 32),
		playerCol = Math.round(object.y / 32);
	return pig.session.level.content[playerRow][playerCol];
}
function hasRightCollision(object) {
	var playerRow = Math.round((object.x + 16) / 32),
		playerCol = Math.round(object.y / 32);
	return pig.session.level.content[playerRow][playerCol];
}


pig-pursuit-master/js/pig.objects.js
/**
 * Pig Pursuit
 *
 * @fileoverview Common objects.
 * @link https://github.com/stevenbenner/pig-pursuit
 * @author Steven Benner
 * @version 0
 * @license Pig Pursuit
 * <https://github.com/stevenbenner/pig-pursuit>
 * Copyright (c) 2012 Steven Benner, http://stevenbenner.com/
 */

pig.objects = {};

pig.objects.sprite = function(url) {
	this.imageUrl = url;
	this.imageObject = null;
	this.imageLoaded = false;
};

pig.objects.sound = function(url) {
	this.url = url;
	this.audioObject = null;
	this.audioLoaded = false;
};

pig.objects.hero = function(leftSprite, rightSprite) {
	this.speed = 256;
	this.jumping = false;
	this.jumpVelocity = 0;
	this.leftSprite = leftSprite;
	this.rightSprite = rightSprite;
	this.health = 256;
	this.isArmed = true;
};

pig.objects.monster = function(type, speed, sprite, health, x, y) {
	this.type = type;
	this.speed = speed;
	this.sprite = sprite;
	this.health = health;
	this.x = x;
	this.y = y;
	this.east = !!Math.round(Math.random() * 1);
};

pig.objects.level = function(width, height) {
	this.backgroundImage = 'images/level1-background.gif';
	this.width = width;
	this.height = height;
	this.isCutscene = false;
	this.events = [];
	this.spawnX = 64;
	this.spawnY = 384;
	this.init = function() {};
	this.monsters = [];
	this.content = []
};

pig.objects.gridElement = function(type, x, y) {
	this.type = type;
	this.x = x;
	this.y = y;
	this.destroy = function() {
		pig.session.level.content[x][y] = null;
	}
}

pig.objects.projectile = function(x, y, angle, speed) {
	this.speed = speed || 900;
	this.x = x;
	this.y = y;
	this.angle = angle;
	this.isHero = !speed;
};

pig.objects.event = function(object, type, startTime, action, duration) {
	this.object = object;
	this.type = type;
	this.startTime = startTime;
	this.action = action;
	this.duration = duration;
	this.inProgress = false;
	this.isComplete = false;
};


pig-pursuit-master/js/pig.gfx.js
/**
 * Pig Pursuit
 *
 * @fileoverview Graphics module.
 * @link https://github.com/stevenbenner/pig-pursuit
 * @author Steven Benner
 * @version 0
 * @license Pig Pursuit
 * <https://github.com/stevenbenner/pig-pursuit>
 * Copyright (c) 2012 Steven Benner, http://stevenbenner.com/
 */

pig.gfx = {};

pig.gfx.sprites = {
	// level sprites
	block: new pig.objects.sprite('images/block.png'),
	blockTop: new pig.objects.sprite('images/block-top.png'),
	blockTopRight: new pig.objects.sprite('images/block-top-right.png'),
	blockTopLeft: new pig.objects.sprite('images/block-top-left.png'),
	blockEarth: new pig.objects.sprite('images/block-earth.png'),
	blockEarthRight: new pig.objects.sprite('images/block-earth-right.png'),
	blockEarthLeft: new pig.objects.sprite('images/block-earth-left.png'),
	door: new pig.objects.sprite('images/door.png'),
	trophy: new pig.objects.sprite('images/trophy.png'),	

	// hero sprites
	heroLeft: new pig.objects.sprite('images/hero-left.png'),
	heroRight: new pig.objects.sprite('images/hero.png'),
	pig: new pig.objects.sprite('images/pig.png'),

	// monster sprites
	mobRoller: new pig.objects.sprite('images/roller.png'),
	mobFlyer: new pig.objects.sprite('images/flyer.png')
};

pig.gfx.createCanvas = function() {
	pig.session.canvas = document.createElement('canvas');
	pig.session.ctx = pig.session.canvas.getContext('2d');
	pig.session.canvas.width = 864;
	pig.session.canvas.height = 512;
	//pig.session.ctx.scale(1.5, 1.5);
	document.getElementById('game').insertBefore(pig.session.canvas, document.getElementById('game').firstChild);
};

pig.gfx.loadImages = function(callback) {
	var bgReady = false;
	pig.session.level.imageObject = new Image();
	pig.session.level.imageObject.onload = function() {
		bgReady = true;
		tryExecCallback();
	}
	pig.session.level.imageObject.src = pig.session.level.backgroundImage;

	for (var type in pig.gfx.sprites) {
		pig.gfx.sprites[type].imageObject = new Image();
		pig.gfx.sprites[type].imageObject.addEventListener('load', (function(type) {
			return function() {
				pig.gfx.sprites[type].imageLoaded = true;
				tryExecCallback();
			}
		})(type), false);
		pig.gfx.sprites[type].imageObject.src = pig.gfx.sprites[type].imageUrl;
	};

	function tryExecCallback() {
		var spritesLoaded = true;
		for (var type in pig.gfx.sprites) {
			if (!pig.gfx.sprites[type].imageLoaded) {
				spritesLoaded = false;
			}
		}
		if (bgReady && spritesLoaded) {
			callback();
		}
	}
};

pig.gfx.renderFrame = function renderFrame() {
	pig.gfx.drawBackground();
	pig.gfx.drawLevelBlocks();
	pig.gfx.drawMonsters();
	pig.gfx.drawHero();
	if (pig.session.hero.isArmed) {
		pig.gfx.drawWeapon();
	}
	pig.gfx.drawProjectiles();
	if (!pig.session.level.isCutscene) {
		pig.session.ctx.drawImage(pig.gfx.drawInfoText(), 16, 16);
		pig.gfx.drawHealthBar();
	}
	if (pig.session.dead) {
		pig.gfx.drawStopText('lol u died');
	} else if (pig.session.paused) {
		pig.gfx.drawStopText('cya soon');
	}
	pig.session.ctx.fillStyle = 'rgb(0, 0, 255)';
	pig.session.ctx.font = 'bold 16px Helvetica';
	pig.session.ctx.textAlign = 'left';
	pig.session.ctx.textBaseline = 'bottom';
	pig.session.ctx.fillText('FPS: ' + ((0.5 + pig.session.fps) << 0), 16, pig.session.canvas.height - 16);
};

pig.gfx.drawBackground = function drawBackground() {
	pig.session.ctx.drawImage(pig.session.level.imageObject, pig.session.xOffset / 10, 0);
};

pig.gfx.drawLevelBlocks = function drawLevelBlocks() {
	var firstColToDraw = Math.abs(Math.ceil(pig.session.xOffset / GRID_SIZE)),
		lastColToDraw = Math.ceil(pig.session.canvas.width / GRID_SIZE) + firstColToDraw,
		x, y, element;
	if (lastColToDraw >= pig.session.level.width) {
		lastColToDraw = pig.session.level.width - 1;
	}
	for (x = firstColToDraw; x <= lastColToDraw; x++) {
		for (y = pig.session.level.height; y--;) {
			element = pig.session.level.content[x][y];
			if (element) {
				pig.session.ctx.drawImage(
					pig.gfx.sprites[element.type].imageObject,
					(x * GRID_SIZE) + pig.session.xOffset,
					y * GRID_SIZE
				);
			}
		}
	}
};

pig.gfx.drawHero = function drawHero() {
	var heroRealX = pig.session.canvas.offsetLeft + pig.session.hero.x + 13 + pig.session.xOffset,
		width;
	pig.session.ctx.drawImage(
		pig.session.mouseX <= heroRealX ? pig.session.hero.leftSprite.imageObject : pig.session.hero.rightSprite.imageObject,
		pig.session.hero.x + pig.session.xOffset,
		pig.session.hero.y
	);
	if (pig.session.hero.message) {
		width = (pig.session.hero.message.length * 8) + 20;
		pig.session.ctx.fillStyle = 'rgb(0,0,0)';
		pig.session.ctx.fillRect(
			pig.session.hero.x + pig.session.xOffset - (width / 2) + 16 - 1,
			pig.session.hero.y - 64 - 1,
			width + 2,
			32 + 2
		);
		pig.session.ctx.fillStyle = 'rgb(255,255,255)';
		pig.session.ctx.fillRect(
			pig.session.hero.x + pig.session.xOffset - (width / 2) + 16,
			pig.session.hero.y - 64,
			width,
			32
		);
		pig.session.ctx.fillStyle = 'rgb(0, 0, 0)';
		pig.session.ctx.font = 'bold 16px Helvetica';
		pig.session.ctx.textAlign = 'center';
		pig.session.ctx.textBaseline = 'top';
		pig.session.ctx.fillText(
			pig.session.hero.message,
			pig.session.hero.x + pig.session.xOffset + 16,
			pig.session.hero.y - 64 + 10
		);
	}
};

pig.gfx.drawWeapon = function drawWeapon() {
	var canvasPosX = pig.session.canvas.offsetLeft,
		canvasPosY = pig.session.canvas.offsetTop,
		heroRealX = canvasPosX + pig.session.hero.x + 13 + pig.session.xOffset,
		heroRealY = canvasPosY + pig.session.hero.y + 23,
		angle = Math.atan2(pig.session.mouseY - heroRealY, pig.session.mouseX - heroRealX),
		startX = pig.session.hero.x + 13 + pig.session.xOffset - 10 * Math.cos(angle),
		startY = pig.session.hero.y + 23 - 10 * Math.sin(angle),
		endX = pig.session.hero.x + 13 + pig.session.xOffset + 24 * Math.cos(angle),
		endY = pig.session.hero.y + 23 + 24 * Math.sin(angle);
	pig.session.ctx.beginPath();
	pig.session.ctx.lineWidth = 4;
	pig.session.ctx.moveTo(startX, startY);
	pig.session.ctx.lineTo(endX, endY);
	pig.session.ctx.stroke();
};

pig.gfx.drawMonsters = function drawMonsters() {
	var i;
	for (i = pig.session.level.monsters.length; i--;) {
		pig.session.ctx.drawImage(
			pig.session.level.monsters[i].sprite.imageObject,
			pig.session.level.monsters[i].x + pig.session.xOffset,
			pig.session.level.monsters[i].y
		);

		if (pig.session.level.monsters[i].message) {
			width = (pig.session.level.monsters[i].message.length * 8) + 20;
			pig.session.ctx.fillStyle = 'rgb(0,0,0)';
			pig.session.ctx.fillRect(
				pig.session.level.monsters[i].x + pig.session.xOffset - (width / 2) + 16 - 1,
				pig.session.level.monsters[i].y - 64 - 1,
				width + 2,
				32 + 2
			);
			pig.session.ctx.fillStyle = 'rgb(255,255,0)';
			pig.session.ctx.fillRect(
				pig.session.level.monsters[i].x + pig.session.xOffset - (width / 2) + 16,
				pig.session.level.monsters[i].y - 64,
				width,
				32
			);
			pig.session.ctx.fillStyle = 'rgb(0, 0, 0)';
			pig.session.ctx.font = 'bold 16px Helvetica';
			pig.session.ctx.textAlign = 'center';
			pig.session.ctx.textBaseline = 'top';
			pig.session.ctx.fillText(
				pig.session.level.monsters[i].message,
				pig.session.level.monsters[i].x + pig.session.xOffset + 16,
				pig.session.level.monsters[i].y - 64 + 10
			);
		}
	}
};

pig.gfx.drawProjectiles = function drawProjectiles() {
	var i;
	for (i = pig.session.projectiles.length; i--;) {
		pig.session.ctx.fillStyle = pig.session.projectiles[i].isHero ? 'rgb(0,100,160)' : 'rgb(123,255,0)';
		pig.session.ctx.fillRect(
			pig.session.projectiles[i].x + pig.session.xOffset,
			pig.session.projectiles[i].y,
			PROJECTILE_SIZE,
			PROJECTILE_SIZE
		);
	}
};

var preRenderText = null;
var lastScore = null;

pig.gfx.drawInfoText = function drawInfoText() {
	if (pig.session.score === lastScore) {
		return preRenderText;
	}
	lastScore = pig.session.score;
var m_canvas = document.createElement('canvas');
m_canvas.width = 128;
m_canvas.height = 64;
var m_context = m_canvas.getContext('2d');
	// draw info text
	m_context.fillStyle = 'rgb(250, 250, 250)';
	m_context.font = 'bold 18px courier new';
	m_context.textAlign = 'left';
	m_context.textBaseline = 'top';
	m_context.fillText('KillCount: ' + pig.session.score, 0, 0);

	preRenderText = m_canvas;
	return preRenderText;
};

pig.gfx.drawHealthBar = function drawHealthBar() {
	var color;
	pig.session.ctx.fillStyle = 'rgb(0,0,0)';
	pig.session.ctx.fillRect(
		pig.session.canvas.width - 267,
		15,
		258,
		18
	);
	if (pig.session.hero.health > 160) {
		color = 'rgb(0,255,0)'
	} else if (pig.session.hero.health > 96) {
		color = 'rgb(255,255,0)'
	} else {
		color = 'rgb(255,77,0)'
	}
	pig.session.ctx.fillStyle = color;
	pig.session.ctx.fillRect(
		pig.session.canvas.width - 266,
		16,
		pig.session.hero.health,
		16
	);
};

pig.gfx.drawStopText = function drawStopText(text) {
	pig.session.ctx.fillStyle = 'rgb(255, 255, 255)';
	pig.session.ctx.font = 'bold 51px courier new';
	pig.session.ctx.textAlign = 'center';
	pig.session.ctx.textBaseline = 'middle';
	pig.session.ctx.lineWidth = 6;
	pig.session.ctx.strokeStyle = '#000000';
	pig.session.ctx.strokeText(text, pig.session.canvas.width / 2, pig.session.canvas.height / 2);
	pig.session.ctx.fillText(text, pig.session.canvas.width / 2, pig.session.canvas.height / 2);
};
